React/state.md

```diff
@@ -1,93 +1,186 @@
-  }
-}
-
-export default NameDisplay;
+// src/components/Student.js
+// ...
+
+const Student = (props) => {
+
+  const [present, setPresent] = useState(false);
+  // Component functions always return JSX
+  return (
+    <div>
+      <h3>{props.fullName}</h3>
+      <ul>
+        <li>Class: C13</li>
+        <li>Birthday: {props.birthday}</li>
+        <li>Email: {props.email}</li>
+      </ul>
+      <button>
+        Mark {present ? 'Absent' : 'Present'}
+      </button>
+    </div>
+  );
+};
 ```
 
-Once we know that the initial state is set, we can use this value within our component class. In this example, we'll use the combination of two state variables to determine whether or not to display a specific property value in our render function.
+Now the component renders with a button that displays "Mark" and either present or absent depending on the current value of `present`.  Try changing the default value of `present` and see the change in the browser.
+
+## Rules with Hooks
+
+There are a few rules to keep in mind with hooks like `useState`.
+
+### 1.  You must call hooks from the top-level of a component.
+
+This means that you cannot put `useState` in a loop like this:
 
 ```javascript
-// NameDisplay.js
-import React from 'react';
-
-class NameDisplay extends React.Component {
-  constructor() {
-    super();
-    this.state = {
-      name: 'Ada',
-      displayName: true
-    };
-  }
-
-  render() {
-    let display = 'Sorry, I don\'t know your name.';
-    if (this.state.displayName) {
-      display = `Hello, ${this.state.name}.`;
-    }
-    return (
-      <section>
-        { display }
-      </section>
-    );
-  }
-}
-
-export default NameDisplay;
+  let i = 0;
+  // Big NO NO!
+  studentList.forEach((student) => {
+    [counter, setCounter] = useState(i);  // will generate an error
+    i += 1;
+  });
 ```
 
-Once the initial state is set, we can then make changes using the `setState` function. (**Note:** `setState` will work even if that state variable was not included in the initial `state` object.) This function will merge this object with the existing state object, overriding any existing values on the same variables. `setState` function calls are very often triggered by events that our users can trigger.
+**Why?**
+
+React depends on the hook functions being called in a specific order when a functional component is run.  It will not allow you to put code into your component which could affect the order in which `useState` is called.  You cannot even put `useState` inside an if statement.
+
+### 2.  You can only call hooks in functional components or other hooks
+
+Creating your own hooks is a more advanced topic, but state is intended to provide a way for a component to have information it remembers and changes internally over time.  So it does not make sense for state to be used outside of a component.  Therefore only call `useState` inside a React component.  Later if you create your own hooks, you can call them there.
```

Comment:
```suggestion
State is intended to provide a way for a component to have information it remembers and changes internally over time, so it does not make sense for state to be used outside of a component.  Therefore only call `useState` inside a React component.  Later if you create your own hooks, you can call them there. But that's a topic we don't need to know about right now.
```
I found it confusing to start by addressing creating your own hooks, so I wanted to suggest removing that bit from the beginning of this line.